File: companion/ZippieCompanionKit/Sources/ZippieCompanionKit/DataBudget.swift
import Foundation

/// A hard ceiling on how much cellular the relay may spend.
///
/// WHY THIS BECAME URGENT. While the relay only ran in the foreground, the
/// user watching the screen WAS the budget - it could not burn a plan
/// unattended because it could not run unattended. On-demand (#2250) removed
/// that limiter: the tunnel now starts itself whenever the phone joins the
/// router's wifi and keeps relaying with the screen off. Per-connection daily
/// and monthly caps are table stakes for anything that relays unattended, and
/// the phase-2 review flagged the absence.
///
/// A BUDGET THAT ONLY WARNS IS NOT A BUDGET. Exceeding it stops the relay.
/// The alternative - carry on and show a badge - is the behaviour that
/// produces a surprise bill, and the phone is the one place in this system
/// where the data is metered and someone else pays for it.
public struct DataBudget: Sendable, Equatable, Codable {
    /// Bytes per day. Zero means unlimited, which is the correct default for a
    /// feature nobody has configured: inventing a cap would silently throttle a
    /// working relay.
    public var dailyBytes: UInt64
    /// Bytes per calendar month. Zero means unlimited.
    public var monthlyBytes: UInt64

    public init(dailyBytes: UInt64 = 0, monthlyBytes: UInt64 = 0) {
        self.dailyBytes = dailyBytes
        self.monthlyBytes = monthlyBytes
    }

    public static let unlimited = DataBudget()
    public var isConfigured: Bool { dailyBytes > 0 || monthlyBytes > 0 }
}

/// What the relay is allowed to do right now.
public enum BudgetVerdict: Equatable, Sendable {
    case allowed
    case dailyExhausted(used: UInt64, limit: UInt64)
    case monthlyExhausted(used: UInt64, limit: UInt64)

    public var isAllowed: Bool { self == .allowed }

    /// Said in plain words, because "relay stopped" with no reason is
    /// indistinguishable from a crash.
    public var reason: String? {
        switch self {
        case .allowed:
            return nil
        case let .dailyExhausted(used, limit):
            return "Daily data cap reached (\(Self.mb(used)) of \(Self.mb(limit)))."
                + " Relaying resumes tomorrow."
        case let .monthlyExhausted(used, limit):
            return "Monthly data cap reached (\(Self.mb(used)) of \(Self.mb(limit)))."
                + " Relaying resumes next month."
        }
    }

    private static func mb(_ b: UInt64) -> String {
        String(format: "%.0f MB", Double(b) / 1_048_576)
    }
}

/// Counts relayed bytes against the budget, and rolls the counters over.
///
/// Counts BOTH directions. A relay that only counted upstream would let the
/// download half of a bonded session run unmetered, which on a phone is the
/// larger half.
public struct BudgetLedger: Sendable, Equatable, Codable {
    public var budget: DataBudget
    public private(set) var dayUsed: UInt64
    public private(set) var monthUsed: UInt64
    /// Day-of-era and month index of the counters, so rollover is a comparison
    /// rather than a timer that has to survive the process being suspended -
    /// which, on a phone, it will not.
    public private(set) var dayStamp: Int
    public private(set) var monthStamp: Int

    public init(budget: DataBudget = .unlimited, now: Date = Date(),
                calendar: Calendar = .current) {
        self.budget = budget
        self.dayUsed = 0
        self.monthUsed = 0
        self.dayStamp = Self.dayIndex(now, calendar)
        self.monthStamp = Self.monthIndex(now, calendar)
    }

    static func dayIndex(_ d: Date, _ c: Calendar) -> Int {
        let comps = c.dateComponents([.year, .month, .day], from: d)
        return (comps.year ?? 0) * 10_000 + (comps.month ?? 0) * 100 + (comps.day ?? 0)
    }

    static func monthIndex(_ d: Date, _ c: Calendar) -> Int {
        let comps = c.dateComponents([.year, .month], from: d)
        return (comps.year ?? 0) * 100 + (comps.month ?? 0)
    }

    /// Rolls the day and month counters if the calendar has moved on.
    /// Idempotent, and safe to call on every datagram.
    public mutating func rollover(now: Date = Date(), calendar: Calendar = .current) {
        let d = Self.dayIndex(now, calendar)
        let m = Self.monthIndex(now, calendar)
        if d != dayStamp {
            dayUsed = 0
            dayStamp = d
        }
        if m != monthStamp {
            monthUsed = 0
            monthStamp = m
        }
    }

    public mutating func record(bytes: UInt64, now: Date = Date(),
                                calendar: Calendar = .current) {
        rollover(now: now, calendar: calendar)
        // Saturating rather than wrapping. A counter that wraps past UInt64.max
        // would reset the budget to zero used, which is the one arithmetic
        // outcome that silently disables the cap.
        dayUsed = dayUsed.addingReportingOverflow(bytes).overflow ? .max : dayUsed + bytes
        monthUsed = monthUsed.addingReportingOverflow(bytes).overflow ? .max : monthUsed + bytes
    }

    public func verdict(now: Date = Date(), calendar: Calendar = .current) -> BudgetVerdict {
        var probe = self
        probe.rollover(now: now, calendar: calendar)
        if probe.budget.monthlyBytes > 0 && probe.monthUsed >= probe.budget.monthlyBytes {
            return .monthlyExhausted(used: probe.monthUsed, limit: probe.budget.monthlyBytes)
        }
        if probe.budget.dailyBytes > 0 && probe.dayUsed >= probe.budget.dailyBytes {
            return .dailyExhausted(used: probe.dayUsed, limit: probe.budget.dailyBytes)
        }
        return .allowed
    }
}

File: companion-android/app/src/main/java/app/zippie/companion/DataBudget.kt
package app.zippie.companion

import java.time.LocalDate

/**
 * A hard ceiling on how much cellular the relay may spend.
 *
 * WHY THIS EXISTS. The relay holds a cellular socket open in a foreground
 * service, so it keeps forwarding with the screen off and the phone in a
 * pocket. Nobody is watching, and the person paying for the plan is often not
 * the person who started it - two Pixels are going into this bond and one of
 * them is not the operator's.
 *
 * A BUDGET THAT ONLY WARNS IS NOT A BUDGET. Exceeding it stops the bytes. The
 * alternative - carry on and show a badge - is the behaviour that produces a
 * surprise bill.
 */
data class DataBudget(
    /** Bytes per calendar day. Zero means unlimited, which is the correct
     *  default for a feature nobody has configured: inventing a cap would
     *  silently throttle a working relay. */
    val dailyBytes: Long = 0,
    /** Bytes per calendar month. Zero means unlimited. */
    val monthlyBytes: Long = 0,
) {
    val isConfigured: Boolean get() = dailyBytes > 0 || monthlyBytes > 0

    companion object {
        val unlimited = DataBudget()
    }
}

/** What the relay is allowed to do right now. */
sealed class BudgetVerdict {
    object Allowed : BudgetVerdict()
    data class DailyExhausted(val used: Long, val limit: Long) : BudgetVerdict()
    data class MonthlyExhausted(val used: Long, val limit: Long) : BudgetVerdict()

    val isAllowed: Boolean get() = this is Allowed

    /** Said in plain words, because "relay stopped" with no reason is
     *  indistinguishable from a crash. */
    val reason: String?
        get() = when (this) {
            is Allowed -> null
            is DailyExhausted ->
                "Daily data cap reached (${mb(used)} of ${mb(limit)}). Relaying resumes tomorrow."
            is MonthlyExhausted ->
                "Monthly data cap reached (${mb(used)} of ${mb(limit)}). Relaying resumes next month."
        }

    private fun mb(b: Long): String = "${b / 1_048_576} MB"
}

/**
 * Counts relayed bytes against the budget and rolls the counters over.
 *
 * Counts BOTH directions. A relay that only counted upstream would leave the
 * download half of a bonded session unmetered, which on a phone is the larger
 * half - the carrier bills for it either way.
 *
 * Synchronised because the two forwarding threads (up and down) both record
 * into it, and a lost increment is a byte that was spent and not counted.
 */
class BudgetLedger private constructor(
    var budget: DataBudget,
    dayUsed: Long,
    monthUsed: Long,
    dayStamp: Int,
    monthStamp: Int,
) {
    var dayUsed: Long = dayUsed
        private set
    var monthUsed: Long = monthUsed
        private set

    /**
     * Which day and month the counters belong to, as a comparable stamp rather
     * than a timer. A timer would have to survive the process being killed,
     * which on a phone it will not; a date comparison survives anything.
     */
    var dayStamp: Int = dayStamp
        private set
    var monthStamp: Int = monthStamp
        private set

    constructor(budget: DataBudget = DataBudget.unlimited, today: LocalDate = LocalDate.now()) :
        this(budget, 0, 0, dayIndex(today), monthIndex(today))

    /** Rolls the day and month counters if the calendar has moved on.
     *  Idempotent, and safe to call on every datagram. */
    @Synchronized
    fun rollover(today: LocalDate = LocalDate.now()) {
        val d = dayIndex(today)
        val m = monthIndex(today)
        if (d != dayStamp) {
            dayUsed = 0
            dayStamp = d
        }
        if (m != monthStamp) {
            monthUsed = 0
            monthStamp = m
        }
    }

    @Synchronized
    fun record(bytes: Long, today: LocalDate = LocalDate.now()) {
        if (bytes <= 0) return
        rollover(today)
        dayUsed = saturatingAdd(dayUsed, bytes)
        monthUsed = saturatingAdd(monthUsed, bytes)
    }

    @Synchronized
    fun verdict(today: LocalDate = LocalDate.now()): BudgetVerdict {
        rollover(today)
        if (budget.monthlyBytes > 0 && monthUsed >= budget.monthlyBytes) {
            return BudgetVerdict.MonthlyExhausted(monthUsed, budget.monthlyBytes)
        }
        if (budget.dailyBytes > 0 && dayUsed >= budget.dailyBytes) {
            return BudgetVerdict.DailyExhausted(dayUsed, budget.dailyBytes)
        }
        return BudgetVerdict.Allowed
    }

    companion object {
        /** Restores counters that outlived the process. Separate from the public
         *  constructor so a restore can never be mistaken for a fresh start:
         *  starting fresh is exactly how a cap gets silently reset. */
        fun restore(
            budget: DataBudget,
            dayUsed: Long,
            monthUsed: Long,
            dayStamp: Int,
            monthStamp: Int,
        ) = BudgetLedger(budget, dayUsed, monthUsed, dayStamp, monthStamp)

        fun dayIndex(d: LocalDate): Int = d.toEpochDay().toInt()

        fun monthIndex(d: LocalDate): Int = d.year * 12 + d.monthValue

        /**
         * Saturating rather than wrapping. Wrapping past Long.MAX_VALUE would
         * hand back a small - or negative - "used" figure, and that is the one
         * arithmetic outcome that silently disables the cap.
         */
        private fun saturatingAdd(a: Long, b: Long): Long {
            val sum = a + b
            return if (sum < a) Long.MAX_VALUE else sum
        }
    }
}

/**
 * The ledger as a single string, for SharedPreferences.
 *
 * Counters must outlive the process or the cap resets every time Android
 * reclaims the app, which turns a monthly budget into a per-launch one. The
 * leading version field exists so that a future format change is REJECTED
 * rather than parsed as counters - a misread field here reads as "nothing spent
 * yet", the failure that costs money.
 */
object BudgetLedgerCodec {
    private const val VERSION = 1

    fun encode(ledger: BudgetLedger): String =
        listOf(VERSION, ledger.dayUsed, ledger.monthUsed, ledger.dayStamp, ledger.monthStamp)
            .joinToString("|")

    /** Null when the string is absent, truncated, from another version, or not
     *  numeric. The caller starts a fresh ledger, which is the only honest
     *  answer when the record cannot be read. */
    fun decode(raw: String?, budget: DataBudget): BudgetLedger? {
        if (raw.isNullOrBlank()) return null
        val parts = raw.split("|")
        if (parts.size != 5) return null
        if (parts[0].toIntOrNull() != VERSION) return null
        val dayUsed = parts[1].toLongOrNull() ?: return null
        val monthUsed = parts[2].toLongOrNull() ?: return null
        val dayStamp = parts[3].toIntOrNull() ?: return null
        val monthStamp = parts[4].toIntOrNull() ?: return null
        if (dayUsed < 0 || monthUsed < 0) return null
        return BudgetLedger.restore(budget, dayUsed, monthUsed, dayStamp, monthStamp)
    }
}
